### ISSUE TITLE
Reproducing the paper results

### ISSUE THREAD

**Issue by vlfom** [DonkeyShot21/UNO#2]
Dear authors,

Thank you for your exciting work and very clean code. I am having trouble reproducing the results mentioned in the paper and would appreciate it if you could help me.

**1. **Reproducing the UNO results from table 4**.**
I was trying to get the scores on the samples of **novel classes from the test split** (Table 4 in the paper).

I have executed the commands for CIFAR10, CIFAR80-20, CIFAR50-50, and used Wandb for logging. However, the results on all datasets did not match the ones that I see in the paper. I took the results from `incremental/unlabel/test/acc`.

| . | Paper (avg/best) | Reproduced (avg/best) |
| -- | -- | -- |
| CIFAR10 | 93.3 / 93.3 | 90.8 / 90.8 |
| CIFAR80-20 | 72.7 / 73.1 | 65.3 / 65.3 |
| CIFAR50-50 | 50.6 / 50.7 | 44.9 / 45.7 |

![image](https://user-images.githubusercontent.com/5049284/132024416-a7750ce9-e606-4569-a8a0-b34a32281ddb.png)

Potential issues:
- I am not using the exact versions of the packages mentioned in your ReadMe, and for that reason, I have run the CIFAR80-20 experiment twice, manually setting the seed (as in RankStats repo), however, I obtained very similar results. I also would not suspect a ~7% difference on CIFAR80-20 just to to the package version.
- I may be using the wrong metric from wandb (I have used `incremental/unlabel/test/acc`). However, if you check my screenshot, for CIFAR80-20 all the other metrics are significantly different anyway (the value close to 72.7/73.1 does not appear anywhere).

**2. **How exactly the RankStats algorithm was evaluated on CIFAR50-50.****

Could you please share if you performed any hyperparameter tuning for the CIFAR50-50 dataset when running the RankStats algorithm on it? I made multiple experiments and my training was very unstable, the algorithm always ends up scoring ~20/17 on known/novel classes.

Thanks a lot for your time.

**Comment by vlfom** [comment]
I have just noticed that the batch size mentioned in the paper is 512, while the one in the ReadMe is 256. I suspect this could be the issue, will test it soon.

**Comment by DonkeyShot21** [comment]
Hi! I have just rerun an experiment with batch size 256 on CIFAR100-80_20 and I got 71.7 for `incremental/unlab/test/acc/avg`, which is very close to the result published in the paper with batch size 512. 
![image](https://user-images.githubusercontent.com/19649581/132037141-cbbb67a0-1de0-419d-8f86-d1034b0f3308.png)

Also if you check the shape of the curve, it looks very different from your curves. I guess you are doing something wrong. It is very likely due to the version of `pytorch-lightning` and/or `lightning-bolts`. Try to use the exact versions I specified in the README.

Regarding ranking statistics, I remember I had some problems with hyperparameters too but in the end it was quite easy to get it running. I am going through my logs now and it seems I used this command for RS+:
```
auto_novel.py --dataset_root ./data/datasets/CIFAR/ --exp_root ./data/experiments/ --warmup_model_dir ./data/experiments/supervised_learning/resnet_rotnet_cifar100-50.pth --lr 0.1 --gamma 0.1 --weight_decay 1e-4 --step_size 340 --batch_size 256 --epochs 400 --rampup_length 300 --rampup_coefficient 25 --num_labeled_classes 50 --num_unlabeled_classes 50 --dataset_name cifar100 --IL --increment_coefficient 0.05 --seed 0 --model_name resnet_IL_cifar100 --mode train --comment RS+-50_50
```
I did not modify their code much so it should just work.

**Comment by vlfom** [comment]
It was indeed an issue specifically with the libraries you mentioned, - I completely overlooked that they affect the optimization, etc.

Thank you for your prompt reply and time @DonkeyShot21.

**Comment by vlfom** [comment]
As a side comment, I noticed that to evaluate your algorithm in the "task-agnostic" setup you generate predictions and solve the assignment problem separately for samples of known and novel classes.

Correct me if I'm wrong, but this might be the wrong way to perform such evaluation because one should have no knowledge on whether the group of samples belongs to known or novel classes, which you assume when performing such evaluation. The way I expected it to be is to use a joint dataloader that contains all the samples mixed and to apply Hungarian on all the resulting predictions at once.

It probably has negligible implications for this work because the classifier is trained well and the assignment problem solver assigns only one logit index to each class. But, as far as I see, assuming some novel class gets confused for the known class the majority of the time (I mean the highest logit it gets comes from the first head), such evaluation would not catch that as an error.

It might be, however, that such evaluation is expected and I just didn't get it from the paper, so just sharing my thoughts here.

**Comment by DonkeyShot21** [comment]
Happy to help! I added a note in the README that warns about package versions.

Regarding the evaluation, I think the procedure I am following is **correct** because I am first concatenating the logits (`preds_inc`) and then I am taking the max of those concatenated logits. By doing this I lose the information about the task. Then in the `compute()` method of the metric class, I compute the best mapping on all classes (and not separately).

**Comment by vlfom** [comment]
The (potential) problem, I believe, is that the `linear_sum_assignment` is computed not for all the data at once.

In the case when one tests on data with novel classes only, assume the model erroneously predicts logits from the first head for all the samples. Then the accuracy should be zero because you explicitly train the first head to predict known classes. However, because you test only on novel classes, the `linear_sum_assignment` could "distribute"/"match" logits from the first head to the novel classes and the accuracy would be above zero. There would be no problem with assigning those first-head logits to novel classes because no "known" images are being fed to `linear_sum_assignment` and those first-head logits are "free" to be assigned to something else.

However, if you would have a mixed dataset, then `linear_sum_assignment` would most likely have assigned those first-head logits to known classes, because they would dominate. And the novel classes would have to be assigned to something else, marking all the first-head predictions for novel classes as errors, and the final accuracy would be much lower.

But, once again, I think it doesn't affect evaluation for balanced and moderately large datasets like CIFAR/ImageNet.

**Comment by DonkeyShot21** [comment]
Ok, now I understand. You are right, this is a potential problem. However, the assignments are quite stable (they are computed on the whole validation set) and, as you said, the potential issue never happens in practice. I remember I tried once to remove the unwanted assignments (the ones that contradicted the labeled head), but the results were exactly the same, while the code was more complicated, so I just removed it. Also, if I remember correctly, in Ranking Statistics they use the same evaluation procedure, so I just sticked to that.
